secure/net/src/network_controller/outbound_handler.rs

```diff
@@ -29,84 +30,126 @@ impl OutboundHandler {
+        // the cost of the mutex when there is no contention
+        let outbound_handlers = mem::take(self.handlers.as_mut());
+
+        // TODO: Consider using multiple tasks for outbound handlers
+        rt.spawn(async move {
+            info!("Starting outbound handler at {}", address.to_string());
+            Self::process_one_outgoing_message(
+                outbound_handlers,
+                &address,
+                inbound_handler.clone(),
+                &mut grpc_clients,
+            )
+            .await;
+            info!("Stopping outbound handler at {}", address.to_string());
+        });
+        Some(stop_signal_tx)
     }
 
-    fn process_one_outgoing_message(
-        outbound_handlers: Arc<Mutex<Vec<(Receiver<Message>, SocketAddr, MessageType)>>>,
-        network_clients: Arc<Mutex<HashMap<SocketAddr, NetworkClient>>>,
+    async fn process_one_outgoing_message(
+        outbound_handlers: Vec<(Receiver<Message>, SocketAddr, MessageType)>,
         socket_addr: &SocketAddr,
         inbound_handler: Arc<Mutex<InboundHandler>>,
+        grpc_clients: &mut HashMap<SocketAddr, RemoteExecutionClientWrapper>,
     ) {
-        let mut select = Select::new();
-        let handlers = outbound_handlers.lock().unwrap();
+        loop {
+            let mut select = Select::new();
+            for (receiver, _, _) in outbound_handlers.iter() {
+                select.recv(receiver);
+            }
 
-        for (receiver, _, _) in handlers.iter() {
-            select.recv(receiver);
-        }
-        let oper = select.select();
-        let index = oper.index();
-        let msg = oper.recv(&handlers[index].0).unwrap();
-        let remote_addr = &handlers[index].1;
-        let message_type = &handlers[index].2;
-        if remote_addr == socket_addr {
-            // If the remote address is the same as the local address, then we are sending a message to ourselves
-            // so we should just pass it to the inbound handler
-            inbound_handler
-                .lock()
-                .unwrap()
-                .send_incoming_message_to_handler(message_type, msg);
-            return;
-        }
-        let mut binding = network_clients.lock().unwrap();
-        let network_client = binding.get_mut(remote_addr).unwrap();
-        let msg = bcs::to_bytes(&NetworkMessage::new(
-            *socket_addr,
-            msg,
-            message_type.clone(),
-        ))
-        .unwrap();
+            let index;
+            let msg;
+            {
+                let oper = select.select();
+                index = oper.index();
+                match oper.recv(&outbound_handlers[index].0) {
+                    Ok(m) => {
+                        msg = m;
+                    },
+                    Err(e) => {
+                        // not necessarily an error, just means a sender closed a channel
+                        trace!(
```

Comment:
Why is this not an error? Silently ignoring error here might mean the transport layer doesn't provide reliable message delivery..

Reply:
This just means one of the senders has closed the send channel. It can happen when the sender_tx channel is dropped after it has done its job.
Also, we should not panic at this level if one the sender dies because many senders subscribe to this outbound task. If a sender_tx channel dies due to an error, then it is application error, not transport layer error.

Reply:
This can mean many things (not necessarily sender dropping).. so I would be hesitant to ignore this and continue.. what is the downside of panic/exiting from this loop when this happens?

Reply:
Ok as discussed, panic'ing on error here

Reply:
@manudhundi - As discussed, instead of continuing here, let's either panic or shut down the handler in case of error?

Reply:
Shutting down the handler. Panic and shutdown would mean the same I suppose as the task just runs this handler

Reply:
In case we want to shut down, we should break out of this loop right?

Reply:
Yes we do break out using the signal from the msg of type "stop_task"